CONTRIBUTING.md

```diff
@@ -456,6 +456,172 @@ checks if contributed code does not have the appropriate license headers.
+these messages to construct a high-level timeline of events leading up to an
+incident.
+
+This level is enabled by default so its target audience is the general
+population of users and administrators. You should use user-facing terminology
+and ensure that messages at this level are self-contained. In general you
+shouldn't log exceptions with stack traces at `INFO` level: if the exception is
+relatively benign then use `DEBUG`, whereas if the user should be notified then
+use `WARN`.
+
+##### `WARN`
+
+This is the next least verbose level, and is also enabled by default. Ideally a
+healthy cluster will emit no `WARN`-level logs, but this is the appropriate
+level for recording events that the cluster administrator should investigate,
+or which indicate a bug. Some production environments require the cluster to
+emit no `WARN`-level logs during acceptance testing, so you must ensure that
+any logs at this level really do indicate a problem that needs addressing.
+
+As with the `INFO` level, you should use user-facing terminology at the `WARN`
+level, and also ensure that messages are self-contained. Strive to make them
+actionable too since you should be logging at this level when the user should
+take some investigative action. Unlike at the `INFO` level, it is often
+appropriate to log an exception, complete with stack trace, at `WARN` level.
+Although the stack trace may not be useful to the user, it may contain
+information that is vital to fully understand the problem. In other cases it
+may be appropriate to log the exception message at `WARN` and only log the full
+exception if the user has enabled `DEBUG` logging:
+
+    if (logger.isDebugEnabled()) {
+        logger.warn("investigate me", exception);
+    } else {
+        logger.warn("investigate me: [{}]", exception.getMessage())
+    }
+
+In a situation where occasional transient failures are expected and handled,
+but a persistent failure requires the user's attemption, consider implementing
+a mechanism to detect that a failure is unacceptably persistent and emit a
+corresponding `WARN` log. For example, it may be helpful to log every tenth
+consecutive failure at `WARN` level, or log at `WARN` if an operation has not
+completed within a certain time limit. This is much more user-friendly than
+failing persistently and silently by default and requiring the user to enable
+`DEBUG` logging to diagnose any problems.
+
+If an exception occurs as a direct result of a request received from a client
+then you should include the exception in the response back to the client rather
+than recording it in the logs at `WARN` level. The person reading the logs is
+usually unable to address any problems caused by faulty client requests, and
+the person running the client is often forbidden from seeing the server logs.
+
+##### `ERROR` and `FATAL`
+
+These are the least verbose levels, but they are not meaningfully different
+from `WARN` within Elasticsearch and you should almost never use them.
```

Comment:
I think it's worth adding a bit more context here - even if this is our desired usage, there are several hundred uses of `.error()`, but just a single non-test use of `.fatal()` currently in the Elasticsearch codebase.

Is `ERROR`-level logging something we should be avoiding in new code? I haven't heard of any such desire, but it's very possible I'm just out of the loop.

Reply:
Perhaps I'm just using [Cunningham's law](https://meta.wikimedia.org/wiki/Cunningham%27s_Law) here :)

I've seen folks get very worried by `ERROR`-level logging emitted by components on the grounds that log4j documents this level as for events which "might still allow the application to continue running" implying that there's a possibility that Elasticsearch actually _won't_ continue running after emitting such a log. There's not zero errors that fit that description, but it's a vanishingly small subset and almost none of the usages of this level fit that interpretation today.

Reply:
Perhaps a discussion to promote back to the wider ES team - I'm certainly not opposed to reducing our usage of `ERROR`-level logs, and a clearer definition of when to use them. I do think there are some uses for it which aren't necessarily fatal but are more severe than a warning. 

For example, it's possible (mostly through misconfiguration nowadays) for ILM to fail to roll over an index, and if that goes "untreated", the index is likely to continue to grow until its shards are big enough to start causing serious problems with the cluster. It's not immediately fatal, but represents a problem that needs to be handled sooner rather than later and using different severity level to indicate that may be useful.

Basically, I'm advocating for EITHER saying `Generally use neither ERROR nor FATAL, and if you see either re-evaluate if it should really be at that severity and adjust it if necessary` OR `Use ERROR for cases fitting these criteria: [...] and use FATAL only for errors which actually kill the process`. But just saying "Don't use `ERROR` or `FATAL`" when we use `ERROR` a bunch in existing code is going to be confusing.

Reply:
Ok I've reworded it to be more permissive around `ERROR`, more highlighting the consequences of using it than saying not to use it ever.

Reply:
We could also bring it back to the team for further discussion, I just think that as written now it should be acceptable.

Reply:
Thanks, I'm good with the new wording for this document. I do think it might be a good idea to solidify this a bit more and re-evaluate the existing logs at some point, but that can be done as time and bandwidth allow.